extensions/stats/plugin.h

```diff
@@ -0,0 +1,405 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <google/protobuf/util/json_util.h>
+#include "absl/container/flat_hash_map.h"
+#include "absl/hash/hash.h"
+#include "absl/strings/str_replace.h"
+#include "extensions/common/context.h"
+#include "extensions/common/node_info.pb.h"
+#include "extensions/stats/config.pb.h"
+
+// WASM_PROLOG
+#ifndef NULL_PLUGIN
+#include "api/wasm/cpp/proxy_wasm_intrinsics.h"
+
+#else  // NULL_PLUGIN
+
+#include "extensions/common/wasm/null/null.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+namespace Null {
+namespace Plugin {
+#endif  // NULL_PLUGIN
+
+// END WASM_PROLOG
+
+namespace Stats {
+
+using StringView = absl::string_view;
+
+constexpr StringView Sep = "#@";
+
+// The following need to be std::strings because the receiver expects a string.
+const std::string unknown = "unknown";
+const std::string vSource = "source";
+const std::string vDest = "destination";
+const std::string vMTLS = "mutual_tls";
+const std::string vNone = "none";
+const std::string vDash = "-";
+
+const std::vector<std::pair<const absl::string_view, std::string>>
+    replacements = {{".", "/"}};
+
+using google::protobuf::util::JsonParseOptions;
+using google::protobuf::util::Status;
+
+#define DBG(dbgsym, ...)                                                    \
+  if ((dbgsym)) {                                                           \
+    logInfo(absl::StrCat(__FILE__, ":", __LINE__, ":", __FUNCTION__, "() ", \
+                         __VA_ARGS__));                                     \
+  }
+
+#define CTXDEBUG(...) DBG(debug_, __VA_ARGS__)
+
+#define ISTIO_DIMENSIONS            \
+  X(reporter)                       \
+  X(source_workload)                \
+  X(source_workload_namespace)      \
+  X(source_principal)               \
+  X(source_app)                     \
+  X(source_version)                 \
+  X(destination_workload)           \
+  X(destination_workload_namespace) \
+  X(destination_principal)          \
+  X(destination_app)                \
+  X(destination_version)            \
+  X(destination_service_host)       \
+  X(destination_service_name)       \
+  X(destination_service_namespace)  \
+  X(request_protocol)               \
+  X(response_code)                  \
+  X(response_flags)                 \
+  X(connection_security_policy)
+
+struct IstioDimensions {
+#define X(name) std::string(name);
+  ISTIO_DIMENSIONS
```

Comment:
I'm not sure the purpose of a macro here. Is there anything else going on here besides defining a bunch of string fields in a struct?

Reply:
Instead of X, it might be better for readability to make it a parameter to another macro, e.g. https://github.com/envoyproxy/envoy/blob/master/include/envoy/http/header_map.h#L264

Reply:
it gives to_string, equality, list_values *and* the struct itself. I have renamed X to something better.

Reply:
I was referring to using macro parameters instead of def/undef cycles.